### ISSUE TITLE
support freeimage plugin instalaltion using pip install imageio[freeimage]

### ISSUE THREAD

**Issue by martinResearch** [imageio/imageio#809]
At the moment the freeimage plugin needs to be installed using either `imageio_download_bin freeimage` or `imageio.plugins.freeimage.download()`. This has to be run after the python environment has been set up with pip.

This is not ideal as one may want to be able to install the plugin when setting up the python environment with pip 
and not later at run time or in a post-installation script. One would also want the executable to be downloaded in the imageio module folder instead of being download in the user local folder. 
In contrast the plugin `gdal` can be installed `pip install imageio[gdal]` and thus is would be great to support `pip install imageio[freeimage]` as well.

**Comment by FirefoxMetzger** [comment]
I agree that it would be nice to have this. If you would like to look into the mechanics of how to implement this, I'd be happy to review.

For general context, we would very much like to turn the freeimage plugin into a dedicated, external package on which ImageIO can depend and that can be installed via `pip install imageio[freeimage]`. This way, we could ship the binary at installation time and rely on pip to manage the download. So far, however, there hasn't been a volunteer to lead this and I have been busy implementing other features. If you are interested, there is definitely room for a contribution here; otherwise we will have to wait until someone steps up and looks into this.

**Comment by martinResearch** [comment]
Would that mean create a new repository https://github.com/imageio/imageio-freeimage (similarly to what is done for opencv with https://github.com/imageio/imageio-ffmpeg) that would simply package the freeimage dll in a python wheel and maybe extract some python code from https://github.com/imageio/imageio/blob/master/imageio/plugins/freeimage.py and https://github.com/imageio/imageio/blob/master/imageio/plugins/_freeimage.py that does not depend on the imageio library in order to have a minimal wrapper around the dll ? I am not sure we are allowed to package the freeimage dll in the python wheels distributed from pypi from a license point of view. 
I would need to get a better idea of how long that could take to code before committing to doing any contribution in that direction.

**Comment by FirefoxMetzger** [comment]
> Would that mean create a new repository ...

Yep, I would create a repository under imageio/imageio-freeimage where the backend would live. I'd also be happy to rely on a freeimage wrapper that is externally maintained, but I am not aware of anything.

> ... that would simply package the freeimage dll in a python wheel and maybe extract some python code and that does not depend on the imageio library in order to have a minimal wrapper around the DLL

More or less, yes; although we never decided on the finer details. @almarklein might have some good ideas on this.

My understanding would be that the [_freeimage.py](https://github.com/imageio/imageio/blob/master/imageio/plugins/_freeimage.py) file moves into a dedicated repo (imageio-freeimage) that is installed together with the binary. This can then be imported by [freeimage.py](https://github.com/imageio/imageio/blob/master/imageio/plugins/freeimage.py) (the current plugin), which can stay in this repo.

In an ideal world, we would also use this chance to add plumbing to build freeimage from source to improve platform support and stay up to date and perhaps update the freeimage plugin into a v3 plugin. However, to keep things pragmatic, a switch into a dedicated repo plus downloading of the binary at install time seems like a good enough start.

>  I am not sure we are allowed to package the freeimage dll in the python wheels distributed from pypi from a license point of view.

That is an interesting point; are you worried about a particular license? 

For freeimage we would use its "FreeImage Public License - Version 1.0", which - unless I am mistaken - has similar terms to LGPL, i.e., we can redistribute the binary but may have to include a copyright notice depending on what we are doing.

> I would need to get a better idea of how long that could take to code before committing to doing any contribution in that direction.

Of course. I guess you are the best person to estimate the time you would need and what is feasible for you. That said, I'm happy to discuss and clarify any questions you may have regarding the plan or direction for something like this.

**Comment by almarklein** [comment]
This is basically a duplicate of #432. We could close that older issue and continue the discussion here. 

> ... that would simply package the freeimage dll in a python wheel and maybe extract some python code and that does not depend on the imageio library in order to have a minimal wrapper around the DLL

Exactly. That package would expose a Pythonic but probably relatively low-level API, and we'd adjust `imageio/plugins/freeimage.py` to use that instead.

The biggest technical challenge would be the CI / packaging. I think it would be good to try and compile freeimage from source, since you'd already be working with GH actions. The cibuildwheel tool might help here. An example is [here](https://github.com/rougier/freetype-py/blob/master/.github/workflows/ci.yml). But just wrapping a pre-build DLL would be fine too.

>>  I am not sure we are allowed to package the freeimage dll in the python wheels distributed from pypi from a license point of view.
>>
> That is an interesting point; are you worried about a particular license?
>
> For freeimage we would use its "FreeImage Public License - Version 1.0", which - unless I am mistaken - has similar terms to LGPL, i.e., we can redistribute the binary but may have to include a copyright notice depending on what we are doing.

No, it's not that simple. For a thread on FreeImage's license, see #114. Fun fact, this issue has been a driver for adding Pillow support. Short story is that the FreeImage license is basically the Mozilla Public license v1, which is not compatible with GPL. I *think* that you could put the DLL's in wheels, but the new `imageio-freeimage` lib would probably have the same FreeImage license. We should add some big warning signs. No package can make it a dependency. So it's only use would be for people using it for themselves.

Reading up on the license issue again, plus the fact that FreeImage is *still* on sourceforge and basically inactive, makes me think that perhaps we should just drop FreeImage ...

**Comment by FirefoxMetzger** [comment]
> No, it's not that simple. [...] but the new imageio-freeimage lib would probably have the same FreeImage license. 

Yes, `imageio-freeimage` would be a derivative work and will have to be GPL or FIPL licensed. ImageIO itself, however, remains BSD.

> We should add some big warning signs. No package can make it a dependency. So it's only use would be for people using it for themselves.

We can do the reverse. Have `imageio-freeimage` depend on ImageIO and on freeimage and also migrate the freeimage.py plugin. This way the license propagation is clear. `imageio` is BSD since it isn't a derivative work of freeimage by virtue of not (optionally) depending on it, freeimage is GPL/FIPL, and imageio-freeimage is GPL/FIPL.

> Reading up on the license issue again, plus the fact that FreeImage is still on sourceforge and basically inactive, makes me think that perhaps we should just drop FreeImage ...

Indeed, that is the easy solution. However, people are still actively using freeimage and its imageio plugin, so we can't just make it disappear without providing alternatives for the use-cases that still exist.

From what I am aware of, this should primarily affect raw formats. Others are mostly be covered by pillow and pyav, but there are a couple other niche formats we would lose support for. Not sure if they are actively used, and we could discover this by adding a deprecation warning and listening to user feedback. That said, I'd like us to decide that we go down this road first before taking such actions.

**Comment by almarklein** [comment]
> We can do the reverse. Have imageio-freeimage depend on ImageIO and on freeimage and also migrate the freeimage.py plugin. This way the license propagation is clear. imageio is BSD since it isn't a derivative work of freeimage by virtue of not (optionally) depending on it, freeimage is GPL/FIPL, and imageio-freeimage is GPL/FIPL.

That makes sense it would indeed make things more clear. Though still need a big warning on the imageio-freeimage page that one should not make it a dependency unless one wants their code to be stuck to GPL/FIPL.

My main worry is that the person(s) that will create and maintain imageio-freeimage are building something that only few people can/will use. An option could be to start really simple, by just copying what's in imageio now. Use the download mechanic, or even require a manual download. Then at least there is *something*.

**Comment by FirefoxMetzger** [comment]
> An option could be to start really simple, by just copying what's in imageio now. Use the download mechanic, or even require a manual download. Then at least there is something.

Yes, that would have been my approach as well. What I don't know is if we also have to remove the config entries in ImageIO that would pick up on `imageio-freeimage` if it exists. No idea if adding config entries for a library is enough to make it count as derivative work (because we are writing code specifically for that library) or if that is fine... GPL at its finest ._.

**Comment by FirefoxMetzger** [comment]
Yes we can, but it looses the ability to auto-import. I.e., you'd have to do something like

```python
import imageio.v3 as iio
import imageio_freeimage
```

This would make `import imageio_freeimage` register the plugin in imageio (assuming we add logic for this) and any code below would pick it up and choose it automatically again.

That is until I figure out how entry points for packages are supposed to work and we can begin to auto-discover/auto-configure plugins that live in external packages. This would be a generic interface so it wouldn't be a derivative work and should be resistant to GPL.
